Compilation errors due to `static_assert(false)` with certain toolset versions (Visual Studio 2022 v143)

## Description
Overload uses `static_assert(false)` inside of `IniFile.inl`, which are problematic when `if constexpr` (compiler hint to execute the branch at compile time) aren't properly execute at compile-time (which happens with certain compilers).

**Reproduced with Visual Studio 2022 (v143)**

## Workaround
Remove all the occurrences of `static_assert(false)`, so [IniFile.inl](https://github.com/Overload-Technologies/Overload/blob/6e4649d0317c48e9ea95bb483c45bfbea5375b80/Sources/Overload/OvTools/include/OvTools/Filesystem/IniFile.inl#L48) line 48, line 94, and line 126

## Expected solution
We should use C++20 concepts to check the type instead of a bunch of `if constexpr`.

## Callstack
```
overload 1.4  Compilation report error。 vs 2022  。   D:\work\Overload\Overload-1.4\Sources\Overload\OvTools\include\OvTools\Filesystem\IniFile.inl(48,18): error C2338: stat
ic_assert failed: 'The given type must be : bool, integral, floating point or string' [D:\work\Overload\Overload-1.4\So
urces\Overload\OvTools\OvTools.vcxproj]
D:\work\Overload\Overload-1.4\Sources\Overload\OvTools\include\OvTools\Filesystem\IniFile.inl(94,19): error C2338: stat
ic_assert failed: 'The given type must be : bool, integral, floating point or string' [D:\work\Overload\Overload-1.4\So
urces\Overload\OvTools\OvTools.vcxproj]
D:\work\Overload\Overload-1.4\Sources\Overload\OvTools\include\OvTools\Filesystem\IniFile.inl(126,19): error C2338: sta
tic_assert failed: 'The given type must be : bool, integral, floating point or std::string' [D:\work\Overload\Overload-
1.4\Sources\Overload\OvTools\OvTools.vcxproj]
```

--- a/Sources/Overload/OvTools/include/OvTools/Filesystem/IniFile.h
+++ b/Sources/Overload/OvTools/include/OvTools/Filesystem/IniFile.h
@@ -9,9 +9,15 @@
 #include <string>
 #include <unordered_map>
 
-
 namespace OvTools::Filesystem
 {
+	template<typename T>
+	concept SupportedIniType =
+		std::same_as<T, bool> ||
+		std::same_as<T, std::string> ||
+		std::integral<T> ||
+		std::floating_point<T>;
+
 	/**
 	* The IniFile class represents a file .ini that stores a set of attributes/values that can get read and written
 	*/
@@ -40,10 +46,10 @@ namespace OvTools::Filesystem
 
 		/**
 		* Return the value attached to the given key
-		* If the key doesn't exist, a default value is returned (0, false, "NULL")
+		* If the key doesn't exist, a default value is returned
 		* @param p_key
 		*/
-		template<typename T>
+		template<SupportedIniType T>
 		T Get(const std::string& p_key);
 
 		/**
@@ -52,7 +58,7 @@ namespace OvTools::Filesystem
 		* @param p_key
 		* @param p_default
 		*/
-		template<typename T>
+		template<SupportedIniType T>
 		T GetOrDefault(const std::string& p_key, T p_default);
 
 		/**
@@ -60,7 +66,7 @@ namespace OvTools::Filesystem
 		* @param p_key
 		* @param p_outValue
 		*/
-		template<typename T>
+		template<SupportedIniType T>
 		bool TryGet(const std::string& p_key, T& p_outValue);
 
 		/**
@@ -68,7 +74,7 @@ namespace OvTools::Filesystem
 		* @param p_key
 		* @param p_value
 		*/
-		template<typename T>
+		template<SupportedIniType T>
 		bool Set(const std::string& p_key, const T& p_value);
 
 		/**
@@ -76,7 +82,7 @@ namespace OvTools::Filesystem
 		* @param p_key
 		* @param p_value
 		*/
-		template<typename T>
+		template<SupportedIniType T>
 		bool Add(const std::string& p_key, const T& p_value);
 
 		/**
@@ -96,24 +102,15 @@ namespace OvTools::Filesystem
 		*/
 		bool IsKeyExisting(const std::string& p_key) const;
 
-		/**
-		* Get the content stored in the ini file as a vector of strings (Each string correspond to an attribute pair : Attribute=Value
-		*/
-		std::vector<std::string> GetFormattedContent() const;
-
 	private:
 		void RegisterPair(const std::string& p_key, const std::string& p_value);
 		void RegisterPair(const AttributePair& p_pair);
 
 		void Load();
 
-		AttributePair	ExtractKeyAndValue(const std::string& p_attributeLine)	const;
-		bool			IsValidLine(const std::string& p_attributeLine)	const;
-		bool			StringToBoolean(const std::string& p_value)			const;
-
 	private:
-		std::string		m_filePath;
-		AttributeMap	m_data;
+		std::string m_filePath;
+		AttributeMap m_data;
 	};
 }
 

--- a/Sources/Overload/OvTools/include/OvTools/Filesystem/IniFile.inl
+++ b/Sources/Overload/OvTools/include/OvTools/Filesystem/IniFile.inl
@@ -12,51 +12,39 @@
 
 namespace OvTools::Filesystem
 {
-	template<typename T>
+	template<SupportedIniType T>
 	inline T IniFile::Get(const std::string& p_key)
 	{
-		if constexpr (std::is_same<bool, T>::value)
+		if (!IsKeyExisting(p_key))
 		{
-			if (!IsKeyExisting(p_key))
-				return false;
-
-			return StringToBoolean(m_data[p_key]);
+			return T{};
 		}
-		else if constexpr (std::is_same<std::string, T>::value)
-		{
-			if (!IsKeyExisting(p_key))
-				return std::string("NULL");
 
+		if constexpr (std::same_as<T, bool>)
+		{
+			return m_data[p_key] == "1" || m_data[p_key] == "T" || m_data[p_key] == "t" || m_data[p_key] == "True" || m_data[p_key] == "true";
+		}
+		else if constexpr (std::same_as<T, std::string>)
+		{
 			return m_data[p_key];
 		}
-		else if constexpr (std::is_integral<T>::value)
+		else if constexpr (std::integral<T>)
 		{
-			if (!IsKeyExisting(p_key))
-				return static_cast<T>(0);
-
 			return static_cast<T>(std::atoi(m_data[p_key].c_str()));
 		}
-		else if constexpr (std::is_floating_point<T>::value)
+		else if constexpr (std::floating_point<T>)
 		{
-			if (!IsKeyExisting(p_key))
-				return static_cast<T>(0.0f);
-
 			return static_cast<T>(std::atof(m_data[p_key].c_str()));
-		}
-		else
-		{
-			static_assert(false, "The given type must be : bool, integral, floating point or string");
-			return T();
 		}
 	}
 
-	template<typename T>
+	template<SupportedIniType T>
 	inline T IniFile::GetOrDefault(const std::string& p_key, T p_default)
 	{
 		return IsKeyExisting(p_key) ? Get<T>(p_key) : p_default;
 	}
 
-	template<typename T>
+	template<SupportedIniType T>
 	inline bool IniFile::TryGet(const std::string& p_key, T& p_outValue)
 	{
 		if (IsKeyExisting(p_key))
@@ -68,30 +56,22 @@ namespace OvTools::Filesystem
 		return false;
 	}
 
-	template<typename T>
+	template<SupportedIniType T>
 	inline bool IniFile::Set(const std::string& p_key, const T& p_value)
 	{
 		if (IsKeyExisting(p_key))
 		{
-			if constexpr (std::is_same<bool, T>::value)
+			if constexpr (std::same_as<T, bool>)
 			{
 				m_data[p_key] = p_value ? "true" : "false";
 			}
-			else if constexpr (std::is_same<std::string, T>::value)
+			else if constexpr (std::same_as<T, std::string>)
 			{
 				m_data[p_key] = p_value;
 			}
-			else if constexpr (std::is_integral<T>::value)
-			{
-				m_data[p_key] = std::to_string(p_value);
-			}
-			else if constexpr (std::is_floating_point<T>::value)
-			{
-				m_data[p_key] = std::to_string(p_value);
-			}
 			else
 			{
-				static_assert(false, "The given type must be : bool, integral, floating point or string");
+				m_data[p_key] = std::to_string(p_value);
 			}
 
 			return true;
@@ -100,8 +80,8 @@ namespace OvTools::Filesystem
 		return false;
 	}
 
-	template<typename T>
-	inline bool IniFile::Add(const std::string & p_key, const T & p_value)
+	template<SupportedIniType T>
+	inline bool IniFile::Add(const std::string& p_key, const T& p_value)
 	{
 		if (!IsKeyExisting(p_key))
 		{
@@ -113,17 +93,9 @@ namespace OvTools::Filesystem
 			{
 				RegisterPair(p_key, p_value);
 			}
-			else if constexpr (std::is_integral<T>::value)
-			{
-				RegisterPair(p_key, std::to_string(p_value));
-			}
-			else if constexpr (std::is_floating_point<T>::value)
-			{
-				RegisterPair(p_key, std::to_string(p_value));
-			}
 			else
 			{
-				static_assert(false, "The given type must be : bool, integral, floating point or std::string");
+				RegisterPair(p_key, std::to_string(p_value));
 			}
 
 			return true;

--- a/Sources/Overload/OvTools/src/OvTools/Filesystem/IniFile.cpp
+++ b/Sources/Overload/OvTools/src/OvTools/Filesystem/IniFile.cpp
@@ -4,10 +4,45 @@
 * @licence: MIT
 */
 
-#include "OvTools/Filesystem/IniFile.h"
-#include "OvTools/Utils/String.h"
+#include <OvTools/Filesystem/IniFile.h>
+#include <OvTools/Utils/String.h>
 
 #include <fstream>
+
+namespace
+{
+	std::pair<std::string, std::string> ExtractKeyAndValue(std::string_view p_line)
+	{
+		std::string key;
+		std::string value;
+
+		std::string* currentBuffer = &key;
+
+		for (const char c : p_line)
+		{
+			if (c == '=')
+				currentBuffer = &value;
+			else
+				currentBuffer->push_back(c);
+		}
+
+		return { key, value };
+	}
+
+	bool IsValidLine(std::string_view p_attributeLine)
+	{
+		if (p_attributeLine.size() == 0)
+			return false;
+
+		if (p_attributeLine[0] == '#' || p_attributeLine[0] == ';' || p_attributeLine[0] == '[')
+			return false;
+
+		if (std::count(p_attributeLine.begin(), p_attributeLine.end(), '=') != 1)
+			return false;
+
+		return true;
+	}
+}
 
 OvTools::Filesystem::IniFile::IniFile(const std::string& p_filePath) : m_filePath(p_filePath)
 {
@@ -51,16 +86,6 @@ void OvTools::Filesystem::IniFile::RegisterPair(const AttributePair& p_pair)
 	m_data.insert(p_pair);
 }
 
-std::vector<std::string> OvTools::Filesystem::IniFile::GetFormattedContent() const
-{
-	std::vector<std::string> result;
-
-	for (const auto&[key, value] : m_data)
-		result.push_back(key + "=" + value);
-
-	return result;
-}
-
 void OvTools::Filesystem::IniFile::Load()
 {
 	std::fstream iniFile;
@@ -90,46 +115,11 @@ void OvTools::Filesystem::IniFile::Rewrite() const
 
 	if (outfile.is_open())
 	{
-		for (const auto&[key, value] : m_data)
+		for (const auto& [key, value] : m_data)
+		{
 			outfile << key << "=" << value << std::endl;
+		}
 	}
 
 	outfile.close();
-}
-
-std::pair<std::string, std::string> OvTools::Filesystem::IniFile::ExtractKeyAndValue(const std::string& p_line) const
-{
-	std::string key;
-	std::string value;
-
-	std::string* currentBuffer = &key;
-
-	for (auto& c : p_line)
-	{
-		if (c == '=')
-			currentBuffer = &value;
-		else
-			currentBuffer->push_back(c);
-	}
-
-	return std::make_pair(key, value);
-}
-
-bool OvTools::Filesystem::IniFile::IsValidLine(const std::string & p_attributeLine) const
-{
-	if (p_attributeLine.size() == 0)
-		return false;
-	
-	if (p_attributeLine[0] == '#' || p_attributeLine[0] == ';' || p_attributeLine[0] == '[')
-		return false;
-	
-	if (std::count(p_attributeLine.begin(), p_attributeLine.end(), '=') != 1)
-		return false;
-
-	return true;
-}
-
-bool OvTools::Filesystem::IniFile::StringToBoolean(const std::string & p_value) const
-{
-	return (p_value == "1" || p_value == "T" || p_value == "t" || p_value == "True" || p_value == "true");
 }

